### PR TITLE
[Core] replace narrow-usage RayWorkerVllm to general WorkerWrapper to reduce code duplication

### PR DESCRIPTION
**Pull request by youkaichao**
Currently, `set_cuda_visible_devices` is one step of worker initialization. It is too specific, and usage is very narrow.

Meanwhile, some information during initialization is essentially lost after initialization, e.g. `local_rank` . We cannot retrieve `local_rank` information from `parallel_state` after initialization. (We can retrieve it from `worker`, but it is difficult to get the worker instance inside many files)

This PR first replaces narrow-usage `set_cuda_visible_devices` to general `update_environment_variables`. I would also like to pass `rank`/`local_rank` etc. information through environment variables, so that we can easily retrieve the information anywhere inside the code, without passing the information all the way from function to function. However, this change is larger, and I would like to hear more opinions.

### CODE REVIEW

File: vllm/utils.py [hunk 1]
```diff
@@ -232,8 +232,14 @@ def get_open_port() -> int:
...
 
-def set_cuda_visible_devices(device_ids: List[int]) -> None:
-    os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(map(str, device_ids))
+def update_environment_variables(envs: Dict[str, str]):
+    for k, v in envs.items():
+        if k in os.environ:
+            warnings.warn(
+                f"Overwriting environment variable {k} "
+                f"from {os.environ[k]} to {v}",
+                stacklevel=2)
```
Comment (zhuohan123) [inline vllm/utils.py:241]: Use logger?

File: tests/distributed/test_pynccl.py [hunk 2]
```diff
@@ -32,8 +33,7 @@ def update_env(fn):
     # so we need to pass the environment variables as arguments
     # and update the environment variables in the function
     def wrapper(env):
-        import os
-        os.environ.update(env)
```
Comment (zhuohan123) [inline tests/distributed/test_pynccl.py:36]: Why don't we just use `os.environ.update`?

Comment (youkaichao) [reply tests/distributed/test_pynccl.py:36]: added warning if some env variables already exist. I think it has better user experience.

File: vllm/utils.py (hunk 1 above)
Comment (youkaichao) [reply vllm/utils.py]: fixed in [1e62614](https://github.com/vllm-project/vllm/pull/4024/commits/1e62614bd4373cba1ea78f76b9ee47f7ae1d937a)

**Comment by esmeetu** [comment]
Can we fix #4029 in this PR by supporting choose `devices` in LLM. Users can pass devices='0,1' to use that two devices.

**Comment by youkaichao** [comment]
After this refactor, `RayGPUExecutor` is only related to GPU with respect to `placement_group`. And we should easily extend it to support distributed inference for other devices.

**Comment by youkaichao** [comment]
TL;DR:

1. support calling `_run_workers` with different args for each worker
2. replace narrow-usage RayWorkerVllm to general WorkerWrapper, which can be used for future executor as well
3. **greatly reduce code duplication**, by adding `use_dummy_driver` arg. e.g. for `get_node_and_gpu_ids`

**Comment by njhill** [comment]
@youkaichao @zhuohan123 any chance we could get https://github.com/vllm-project/vllm/pull/3466 merged first? I have been rebasing it for a couple of months now.

Looks like it needs another small rebase, I can do that shortly.

File: vllm/executor/ray_gpu_executor.py [hunk 3]
```diff
@@ -98,13 +97,22 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
...
+                worker_class_name="Worker",
+            )
 
             worker_ip = ray.get(worker.get_node_ip.remote())
             if worker_ip == driver_ip and self.driver_dummy_worker is None:
                 # If the worker is on the same node as the driver, we use it
                 # as the resource holder for the driver process.
                 self.driver_dummy_worker = worker
+                self.driver_worker = RayWorkerWrapper(
+                    init_cached_hf_modules=self.model_config.trust_remote_code,
```
Comment (cadedaniel) [inline vllm/executor/ray_gpu_executor.py:112]: nit: if we pass `trust_remote_code` directly, the implementation itself can decide whether it needs to do `init_cached_hf_modules`. otherwise since the decision is not "owned" by the wrapper, it becomes more coupled with higher layer. (I realize this isn't your addition, OK to ignore)

File: vllm/worker/worker_base.py [hunk 4]
```diff
@@ -81,3 +87,54 @@ def remove_lora(self, lora_id: int) -> bool:
...
+        self.init_cached_hf_modules = init_cached_hf_modules
+        self.worker = None
+
+    def update_environment_variables(self, envs: Dict[str, str]) -> None:
+        key = 'CUDA_VISIBLE_DEVICES'
+        if key in envs and key in os.environ:
+            # overwriting CUDA_VISIBLE_DEVICES is desired behavior
+            # suppress the warning in `update_environment_variables`
+            del os.environ[key]
+        update_environment_variables(envs)
```
Comment (cadedaniel) [inline vllm/worker/worker_base.py:115]: At a high-level, we should discourage passing configuration / data via environment variables because it allows for a lot of complexity that isn't subject to documentation / visibility / common abstractions that a config object would be subject to. Basically, we don't want to give the ability for external contributors to add additional configuration space very easily. An example of this is the Ray DAG vLLM integration, which is very hidden because it's not in a config object. https://github.com/vllm-project/vllm/blob/e95cd879598b834f85e70ebcd23db316ae430540/vllm/executor/ray_gpu_executor.py#L24-L27

Another example is that `CUDA_VISIBLE_DEVICES` semantics may only work for a subset of backends; the way each backend configures which devices to use can be different. Then we have a cambrian explosion of env vars to manage devices..

Alternatives:
* Allow a whitelist of env vars that we're confident apply for all vLLM backends, e.g. MPI's `RANK`/`LOCAL_RANK`/`WORLD_SIZE`/etc
* Take standard vLLM config and allow each backend to interpret / apply it (e.g. `set_visible_devices` API instead of `update_environment_variables`)

File: vllm/worker/worker_base.py [hunk 5]
```diff
@@ -81,3 +87,54 @@ def remove_lora(self, lora_id: int) -> bool:
...
+            return executor(*args, **kwargs)
+        except Exception as e:
+            # if the driver worker also execute methods,
+            # exceptions in the rest worker may cause deadlock in rpc like ray
+            # see https://github.com/vllm-project/vllm/issues/3455
+            # print the error and inform the user to solve the error
+            msg = (f"Error executing method {method}. "
+                   "This might cause deadlock in distributed execution.")
+            logger.exception(msg)
+            raise e
```
Comment (cadedaniel) [inline vllm/worker/worker_base.py:140]: Until we have another example of this happening, this should live in the Ray-specific worker wrapper

File: vllm/utils.py [hunk 6]
```diff
@@ -232,8 +232,12 @@ def get_open_port() -> int:
             return s.getsockname()[1]
 
 
-def set_cuda_visible_devices(device_ids: List[int]) -> None:
-    os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(map(str, device_ids))
+def update_environment_variables(envs: Dict[str, str]):
+    for k, v in envs.items():
+        if k in os.environ:
+            logger.warning(f"Overwriting environment variable {k} "
+                           f"from {os.environ[k]} to {v}")
```
Comment (cadedaniel) [inline vllm/utils.py:239]: nit: do `{os.environ[k]=}` and `{v=}` so whitespace is obvious

File: vllm/worker/worker_base.py [hunk 7]
```diff
@@ -81,3 +87,54 @@ def remove_lora(self, lora_id: int) -> bool:
 
     def list_loras(self) -> List[int]:
         raise ValueError(f"{type(self)} does not support LoRA")
+
+
+class WorkerWrapperBase:
```
Comment (cadedaniel) [inline vllm/worker/worker_base.py:92]: We should strive for composition over inheritance to limit the coupling of the implementations to the interface; otherwise we invite undesired coupling which makes some changes difficult.

File: vllm/worker/worker_base.py [hunk 8]
```diff
@@ -81,3 +87,54 @@ def remove_lora(self, lora_id: int) -> bool:
...
+
+    def update_environment_variables(self, envs: Dict[str, str]) -> None:
+        key = 'CUDA_VISIBLE_DEVICES'
+        if key in envs and key in os.environ:
+            # overwriting CUDA_VISIBLE_DEVICES is desired behavior
+            # suppress the warning in `update_environment_variables`
+            del os.environ[key]
+        update_environment_variables(envs)
+
+    def init_worker(self, *args, **kwargs):
```
Comment (cadedaniel) [inline vllm/worker/worker_base.py:117]: Can we add some docstrings?

File: vllm/executor/ray_gpu_executor.py (hunk 3 above)
Comment (youkaichao) [reply vllm/executor/ray_gpu_executor.py]: fixed in [74deb44](https://github.com/vllm-project/vllm/pull/4024/commits/74deb444c1a694b940720806e8b318da562273f5).

File: vllm/worker/worker_base.py (hunk 5 above)
Comment (youkaichao) [reply vllm/worker/worker_base.py:136]: This is in general the case for rpc framework, where we have the driver worker both running models and check health/get return result from other workers.

File: vllm/utils.py (hunk 6 above)
Comment (youkaichao) [reply vllm/utils.py]: I add single quotes to make it clear. `{os.environ[k]=}` will print `os.environ[k]` to users, which is confusing and not informative.

File: vllm/worker/worker_base.py (hunk 8 above)
Comment (youkaichao) [reply vllm/worker/worker_base.py:116]: fixed in [21be004](https://github.com/vllm-project/vllm/pull/4024/commits/21be00449f892f6480b95e54d2cd94db1c46d720).

File: vllm/worker/worker_base.py (hunk 7 above)
Comment (youkaichao) [reply vllm/worker/worker_base.py:93]: I don't get it. Can you elaborate on this?

File: vllm/worker/worker_base.py (hunk 5 above)
Comment (cadedaniel) [reply vllm/worker/worker_base.py:137]: OK makes sense!

File: vllm/worker/worker_base.py (hunk 8 above)
Comment (cadedaniel) [reply vllm/worker/worker_base.py:116]: This is a comment, docstrings look [like this](https://chat.openai.com/share/6cee29eb-f90b-4ae6-ac17-1464d24470bc) (this degree of formality is overkill for this method..)

Python doc tooling will then take the first statement (if it's a string) and expose it for auto doc generation

example:
https://github.com/vllm-project/vllm/blob/d150e4f89f9da5b093511227449ad940a9a82a52/vllm/core/block/cpu_gpu_block_allocator.py#L155-L167

File: vllm/worker/worker_base.py (hunk 7 above)
Comment (cadedaniel) [reply vllm/worker/worker_base.py:93]: The thinking goes like this:
1. The purpose of an interface is to minimally couple different implementations.
2. Different implementations may have common logic, and this common logic can be put in the abstract class
3. It is good to avoid this because it violates (1); the coupling goes beyond the minimal interface definition
4. The alternative to inheritance is to have utility classes / functions and compose their usage within the different implementations. [example](https://chat.openai.com/share/6ff20436-023f-4ff6-9f2a-8925cb268d3d)

It is not a hard rule as it's impossible to completely decouple different components but a good design principle.

I'll still approve this PR without it; however the benefit is that each worker implementation is more cleanly separated from each other, making future changes easier.

File: vllm/worker/worker_base.py (hunk 4 above)
Comment (cadedaniel) [reply vllm/worker/worker_base.py:114]: Agree! and I won't block the PR on it -- but to raise vLLM quality as we have more backends, we should be careful with global state https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil

OK to have a small set of allowed env vars, but it's a bad idea to support arbitrary ones at interface level. fine to improve later.

File: vllm/worker/cpu_worker.py [hunk 9]
```diff
@@ -138,6 +138,10 @@ def __init__(
         self.is_driver_worker = is_driver_worker
         if self.is_driver_worker:
             assert self.rank == 0, "The driver worker must have rank 0."
+        init_cached_hf_modules = self.model_config.trust_remote_code
+        if init_cached_hf_modules:
+            from transformers.dynamic_module_utils import init_hf_modules
+            init_hf_modules()
```
Comment (cadedaniel) [inline vllm/worker/cpu_worker.py:144]: nit: good to put this new common logic into a util method, with docstring

**Review by cadedaniel — APPROVED** [review]
some nits and suggestions, otherwise LGTM

File: vllm/worker/cpu_worker.py (hunk 9 above)
Comment (youkaichao) [reply vllm/worker/cpu_worker.py]: fixed in [2509db4](https://github.com/vllm-project/vllm/pull/4024/commits/2509db4e62c44192198de1b61c9eda8dd75628a6)

**Comment by youkaichao** [comment]
Thanks for the review!

> OK to have a small set of allowed env vars, but it's a bad idea to support arbitrary ones at interface level. fine to improve later.

I'm thinking of adding a `vllm/flags.py` to hold all the global flags in a central place. Will do that later.

> the benefit is that each worker implementation is more cleanly separated from each other, making future changes easier

I didn't accept the change, because `self.worker_module_name` and `self.worker_class_name` are how `WorkerWrapperBase` works. It would sound very strange to put this code into utils.
